### PR TITLE
[RFR] Init api 

### PR DESCRIPTION
**Pull request by HALLERPierre**
# TODO
- [x] Add node
- [x] Add express & socket.io
- [x] Add postgres
- [x] Create game table
- [x] Linter
- [x] Makefile
- [x] Test (jest)
- [x] Travis

### CODE REVIEW

File: api/package.json [hunk 1]
```diff
@@ -1,15 +1,25 @@
...
   "main": "index.js",
   "scripts": {
     "start": "node app.js",
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "jest",
+    "lint": "eslint",
+    "init-db": "node scripts/initDb.js"
   },
   "author": "",
   "license": "ISC",
```
Comment (Kmaschta) [inline api/package.json]: Nope, it's not licensed ISC

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,26 @@
+.PHONY: help install run stop test
```
Comment (zyhou) [inline Makefile]: run?

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+	docker-compose up -d
+
+stop: ## Stop the server
+	docker-compose down
+
+test-api:
+	docker-compose run --rm \
+		api npm run test
+
+test: ## Test the go code and php code
```
Comment (zyhou) [inline Makefile]: Interesting comment

File: api/.env [hunk 4]
```diff
@@ -0,0 +1,6 @@
+POSTGRES_URL=postgres
```
Comment (zyhou) [inline api/.env]: Don't commit .env file, use .env.dist. `cp -n .env.dist .env`

File: api/.eslintrc.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
+module.exports = {
+  env: {
+    browser: true,
+    commonjs: true,
+    es6: true,
+    jest: true,
+  },
+  extends: 'airbnb-base',
```
Comment (zyhou) [inline api/.eslintrc.js:8]: We don't use aribnb config anymore, use prettier instead

Comment (zyhou) [reply api/.eslintrc.js:8]: You can keep aribnb

File: api/app.js [hunk 6]
```diff
@@ -1,10 +1,17 @@
...
-  res.send('Hello World!')
-})
+const listenerSocketIO = (socket) => {
+  socket.on('new-game', () => createNewGame().then(game => socket.emit('new-game-reply', game)));
+};
 
-app.listen(3000, function () {
-  console.log('Example app listening on port 3000!')
-})
+app.get('/', (req, res) => {
```
Comment (zyhou) [inline api/app.js]: You can remove the rest route

File: api/app.js [hunk 7]
```diff
@@ -1,10 +1,17 @@
...
+const http = require('http').createServer(app);
+const io = require('socket.io')(http);
+const { PORT } = require('./constants/api');
+const createNewGame = require('./app/createNewGame');
 
-app.get('/', function (req, res) {
-  res.send('Hello World!')
-})
+const listenerSocketIO = (socket) => {
+  socket.on('new-game', () => createNewGame().then(game => socket.emit('new-game-reply', game)));
```
Comment (zyhou) [inline api/app.js]: Use constant and async/await

File: api/app/getGame.js [hunk 8]
```diff
@@ -0,0 +1,5 @@
+const { get } = require('../database/game');
```
Comment (zyhou) [inline api/app/getGame.js]: useless file no ?

File: api/database/game.js [hunk 9]
```diff
@@ -0,0 +1,15 @@
+const db = require('./index');
+const { toEntity } = require('../models/game');
+
+const TABLE_NAME = 'public.game';
+
+const get = id => db.query(`SELECT * FROM ${TABLE_NAME} WHERE id = $1`, [id]).then(res => toEntity(res.rows[0]));
```
Comment (zyhou) [inline api/database/game.js:6]: async await

exemples :
```js
const get = async(id) => {
   const { rows } = awaitqb.query(...);
  return toEntity(rows[0])
};
```

File: api/database/game.js [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+const save = game => db
+  .query({
```
Comment (zyhou) [inline api/database/game.js:9]: Why you use object param ? And not from above `.query(queryString, queryArgs)

File: api/jest.config.js [hunk 11]
```diff
@@ -0,0 +1,185 @@
+// For a detailed explanation regarding each configuration property, visit:
```
Comment (zyhou) [inline api/jest.config.js:1]: to much comment, clean this file

File: api/package.json [hunk 12]
```diff
@@ -1,15 +1,25 @@
...
-  "description": "",
+  "author": "Pierre",
+  "description": "Quixo game with node api & react-native app",
   "main": "index.js",
   "scripts": {
     "start": "node app.js",
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "jest",
+    "lint": "eslint",
+    "init-db": "node scripts/initDb.js"
```
Comment (zyhou) [inline api/package.json]: no need you already use `docker-entrypoint-initdb.d`

File: api/models/game.js [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+  if (key === 'currentPlayer') {
+    newKey = 'current_player';
+  }
+  if (key === 'selectedCube') {
+    newKey = 'selected_cube';
+  }
+  if (key === 'board') {
+    val = JSON.stringify(game[key]);
+  }
+  return Object.assign({}, newGame, { [newKey]: val });
```
Comment (zyhou) [inline api/models/game.js]: use spread operator (return {...newGame,[newKey]: val }

File: api/models/game.js [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+
+const toEntity = game => Object.keys(game).reduce((newGame, key) => {
+  let newKey = key;
+  if (key === 'current_player') {
+    newKey = 'currentPlayer';
+  }
+  if (key === 'selected_cube') {
+    newKey = 'selectedCube';
+  }
+  return Object.assign({}, newGame, { [newKey]: game[key] });
```
Comment (zyhou) [inline api/models/game.js]: same

File: api/constants/api.js [hunk 15]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  PORT: process.env.API_PORT || 3000,
```
Comment (zyhou) [inline api/constants/api.js:2]: On all projet at Marmelab, we use convict and dotenv

**Review by zyhou — CHANGES_REQUESTED** [review]

**Review by zyhou — APPROVED** [review]

**Pull request merged**